experimental/oak_async/src/grpc.rs

```diff
@@ -0,0 +1,146 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! gRPC related types for asynchronous processing.
+
+use log::error;
+use oak::{
+    grpc::{ChannelResponseWriter, Code, WriteMode},
+    OakError,
+};
+
+/// Writes exactly one message on a gRPC response channel.
+///
+/// This type is used for gRPC methods where the return type is non-streaming. Instances must be
+/// consumed using `send`. In exceptional circumstances the `into_inner` method may be used to
+/// access the underlying `ChannelResponseWriter`.
+///
+/// If neither `send` nor `into_inner` is called, an `UNKNOWN` error will be written as the response
+/// upon `Drop`.
+#[must_use = "exactly one message should be sent. Use into_inner for advanced use cases"]
+pub struct OneshotWriter<T> {
+    writer: ChannelResponseWriter,
+    _msg: core::marker::PhantomData<T>,
+}
+
+impl<T: ::prost::Message> OneshotWriter<T> {
+    // Only for use by the Oak Service Generator.
+    #[doc(hidden)]
+    pub fn new(writer: ChannelResponseWriter) -> OneshotWriter<T> {
+        OneshotWriter {
+            writer,
+            _msg: core::marker::PhantomData,
+        }
+    }
+
+    /// Send the response, automatically closing the channel.
+    pub fn send(self, msg: &T) -> Result<(), OakError> {
+        self.writer.write(msg, WriteMode::Close)?;
+        // Prevents running the destructor
+        core::mem::forget(self);
+        Ok(())
+    }
+
+    /// Consumes this writer, returning a `ChannelResponseWriter` in its place.
+    pub fn into_inner(self) -> ChannelResponseWriter {
+        into_inner(self, |s| &mut s.writer)
+    }
+}
+
+impl<T> Drop for OneshotWriter<T> {
+    /// Close the channel with an `Unknown` status.
+    fn drop(&mut self) {
+        // TODO(wildarch) or should we panic (after setting the error)?
```

Comment:
@tiziano88 and @daviddrysdale I would like to know what your opinion is on the correct behaviour here. I'm tempted to make this `panic!`.

At the same time it seems like panic-on-drop is frowned upon:
- If the user code panics while a `OneshotWriter` is in scope, this will also invoke `drop()` and panic while panicking, which can cause the process to abort.
- The user has little control over when and where the writer is dropped, so a panic can seem a bit 'random'

At the same time the user would already be ignoring all documentation and a `must_use` annotation before ever triggering this panic, so it seems reasonable to panic in that case.

Reply:
Can we just avoid the problem by making `writer` an `Option<ChannelResponseWriter>`? Something like:

```Rust
pub struct OneshotWriter<T> {
    writer: Option<ChannelResponseWriter>,
    _msg: core::marker::PhantomData<T>,
}

impl<T: ::prost::Message> OneshotWriter<T> {
    // Only for use by the Oak Service Generator.
    #[doc(hidden)]
    pub fn new(writer: ChannelResponseWriter) -> OneshotWriter<T> {
        OneshotWriter {
            writer: Some(writer),
            _msg: core::marker::PhantomData,
        }
    }

    /// Send the response, automatically closing the channel.
    pub fn send(self, msg: &T) -> Result<(), OakError> {
        self.into_inner().write(msg, WriteMode::Close)
    }

    /// Consumes this writer, returning a `ChannelResponseWriter` in its place.
    pub fn into_inner(mut self) -> ChannelResponseWriter {
        std::mem::replace(&mut self.writer, None).unwrap()
    }
}

impl<T> Drop for OneshotWriter<T> {
    /// Close the channel with status that an `Unknown` status.
    fn drop(&mut self) {
        if let Some(writer) = &self.writer {
            if let Err(e) = writer.close(Err(oak::grpc::build_status(Code::Unknown, ""))) {
                error!("Failed to close OneshotWriter in Drop: {}", e);
            }
        }
    }
}
```